pkg/kademlia/routing_integration_test.go

```diff
@@ -0,0 +1,458 @@
+	require.Equal(t, cacheSize, table.CacheSize())
+
+	nodes, err := table.FindNear(PadID("20", "0"), 3)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(nodes))
+}
+
+func TestTableBasic_Routing(t *testing.T)     { testTableBasic(t, newRouting) }
+func TestTableBasic_TestRouting(t *testing.T) { testTableBasic(t, newTestRouting) }
+func testTableBasic(t *testing.T, routingCtor routingCtor) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	table, close := routingCtor(PadID("5555", "5"), 5, 3, 0)
+	defer close()
+
+	err := table.ConnectionSuccess(Node(PadID("5556", "5"), "address:1"))
+	require.NoError(t, err)
+
+	nodes, err := table.FindNear(PadID("20", "0"), 3)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(nodes))
+	require.Equal(t, PadID("5556", "5"), nodes[0].Id)
+	require.Equal(t, "address:1", nodes[0].Address.Address)
+}
+
+func TestNoSelf_Routing(t *testing.T)     { testNoSelf(t, newRouting) }
+func TestNoSelf_TestRouting(t *testing.T) { testNoSelf(t, newTestRouting) }
+func testNoSelf(t *testing.T, routingCtor routingCtor) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	table, close := routingCtor(PadID("55", "5"), 5, 3, 0)
+	defer close()
+	err := table.ConnectionSuccess(Node(PadID("55", "5"), "address:2"))
+	require.NoError(t, err)
+
+	nodes, err := table.FindNear(PadID("20", "0"), 3)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(nodes))
+}
+
+func TestSplits_Routing(t *testing.T)     { testSplits(t, newRouting) }
+func TestSplits_TestRouting(t *testing.T) { testSplits(t, newTestRouting) }
+func testSplits(t *testing.T, routingCtor routingCtor) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	table, close := routingCtor(PadID("55", "5"), 5, 2, 0)
+	defer close()
+
+	for _, prefix2 := range "08" {
+		for _, prefix1 := range "a69c23f1d7eb5408" {
+			require.NoError(t, table.ConnectionSuccess(
+				NodeFromPrefix(string([]rune{prefix1, prefix2}), "0")))
+		}
+	}
+
+	// we just put 32 nodes into the table. the bucket with a differing first
+	// bit should be full with 5 nodes. the bucket with the same first bit and
+	// differing second bit should be full with 5 nodes. the bucket with the
+	// same first two bits and differing third bit should not be full and have
+	// 4 nodes (60..., 68..., 70..., 78...). the bucket with the same first
+	// three bits should also not be full and have 4 nodes
+	// (40..., 48..., 50..., 58...). So we should be able to get no more than
+	// 18 nodes back
+
+	nodes, err := table.FindNear(PadID("55", "5"), 19)
+	require.NoError(t, err)
+	requireNodesEqual(t, []*pb.Node{
+		// bucket 010 (same first three bits)
+		NodeFromPrefix("50", "0"), NodeFromPrefix("58", "0"),
+		NodeFromPrefix("40", "0"), NodeFromPrefix("48", "0"),
+
+		// bucket 011 (same first two bits)
+		NodeFromPrefix("70", "0"), NodeFromPrefix("78", "0"),
+		NodeFromPrefix("60", "0"), NodeFromPrefix("68", "0"),
+
+		// bucket 00 (same first bit)
+		NodeFromPrefix("10", "0"),
+		NodeFromPrefix("00", "0"),
```

Comment:
Failing at NodeFromPrefix("00", "0") in TestSplits_Routing. Expects 1111111111111111111111111111111112m1s9K which represents an empty Node ID. Getting 1N96YXFr9Fc66JCJ7JdEzs2AxchSCHR2JptvQNysgkaDduN6LF instead.